gubernator/github/classifier.py

```diff
@@ -0,0 +1,152 @@
+    '''
+    merged = {}
+    labels = []
+    comments = {}  # comment_id : comment
+    commit_change_time = None
+    for event in events:
+        body = json.loads(event.body)
+        if 'issue' in body:
+            merged.update(body['issue'])
+            # issues come with labels, so we can update here
+            labels = body['issue']['labels']
+        elif 'pull_request' in body:
+            merged.update(body['pull_request'])
+            # pull_requests don't include their labels :(
+        action = body.get('action')
+        sender = body.get('sender', {}).get('login')
+        if event.event in ('issue_comment', 'pull_request_review_comment'):
+            comment_id = body['comment']['id']
+            if action == 'deleted':
+                comments.pop(comment_id, None)
+            else:
+                comments[comment_id] = body['comment']
+        if event.event == 'pull_request':
+            # Pull request label events don't come with a full label set.
+            # Track them explicitly here.
+            if action in ('opened', 'reopened', 'synchronize'):
+                commit_change_time = body['pull_request']['updated_at']
+            try:
+                if action in ('labeled', 'unlabeled') and 'label' not in body:
+                    logging.warning('label event with no labels (multiple changes?)')
+                elif action == 'labeled':
+                    label = body['label']
+                    if label not in labels:
+                        labels.append(label)
+                elif action == 'unlabeled':
+                    label = body['label']
+                    if label in labels:
+                        labels.remove(label)
+            except:
+                logging.exception('??? %r', body)
+                raise
+
+    is_pr = 'head' in merged or 'pull_request' in merged
+    is_open = merged['state'] != 'closed'
+    author = merged['user']['login']
+    assignees = sorted(assignee['login'] for assignee in merged['assignees'])
+    involved = [author] + assignees
+    labels = {label['name']: label['color'] for label in labels}
+    comments = [
+            {
+                'author': c['user']['login'],
+                'comment': c['body'],
+                'timestamp': c['created_at']
+            }
+            for c in sorted(comments.values(), key=lambda c: c['created_at'])
+    ]
+
+    payload = {
+        'author': author,
+        'assignees': assignees,
+        'title': merged['title'],
+        'labels': labels,
+    }
+
+    if is_pr:
+        if is_open:
+            payload['needs_rebase'] = 'needs-rebase' in labels or merged.get('mergeable') == 'false'
+        payload['additions'] = merged.get('additions', 0)
+        payload['deletions'] = merged.get('deletions', 0)
+
+    payload['attn'] = calculate_attention(is_pr, is_open, payload, comments, commit_change_time)
+
+    return is_pr, is_open, involved, payload, events[-1].timestamp
```

Comment:
Might be worth packaging this up into a namedtuple or something along those lines.